### PR TITLE
Support GCS in rikai IO

### PR DESCRIPTION
**Pull request by eddyxu**
Support Google cloud storage in Rikai IO for #20

### CODE REVIEW

**Comment by eddyxu** [comment]
Have tested manually like

```python
from rikai.io import copy
from rikai.types.vision import Image

copy("local.png", "gs://my_bucket/")
copy("gs://my_bucket/local.png", "gs://my_bucket/local_copy.png")

print(Image("gs://my_bucket/local.png").to_numpy())
```

We dont have the infra to test this in rikai yet.

File: python/setup.py [hunk 1]
```diff
@@ -7,8 +7,7 @@
     exec(fh.read(), about)
 
 with open(
-    pathlib.Path(__file__).absolute().parent.parent / "README.md",
-    "r",
+    pathlib.Path(__file__).absolute().parent.parent / "README.md", "r"
```
Comment (eddyxu) [inline python/setup.py:10]: This was re-formatted by `black`

File: python/rikai/io.py [hunk 2]
```diff
@@ -76,17 +109,19 @@ def copy(source: str, dest: str) -> str:
...
 
     # TODO: find better i/o utilis to copy between filesystems
-    filesystem, dest_path = fs.FileSystem.from_uri(dest)
-    with filesystem.open_output_stream(dest_path) as out_stream:
-        src_fs, src_path = fs.FileSystem.from_uri(source)
-        with src_fs.open_input_stream(src_path) as in_stream:
+    with _open_output_stream(dest) as out_stream:
+        with _open_input_stream(source) as in_stream:
             while True:
                 buf = in_stream.read(_BUFSIZE)
```
Comment (chunyang) [inline python/rikai/io.py:126]: Can probably use `shutil` here instead of implementing it ourselves.

File: python/rikai/io.py [hunk 3]
```diff
@@ -49,6 +49,39 @@ def _normalize_uri(uri: str) -> str:
     ).geturl()
 
 
+def _gcsfs():
```
Comment (chunyang) [inline python/rikai/io.py:52]: Use `@functools.cache` instead of maintaining the `_GCSFS` global variable?

File: python/rikai/io.py [hunk 4]
```diff
@@ -124,6 +159,8 @@ def open_uri(uri: Union[str, Path], mode="rb") -> BinaryIO:
     elif parsed_uri.scheme in ("http", "https"):
         resp = requests.get(uri)
         return BytesIO(resp.content)
+    elif parsed_uri.scheme == "gs":
+        return _gcsfs().open(uri, mode=mode)
```
Comment (chunyang) [inline python/rikai/io.py:163]: I think the type signature is wrong? Depending on `mode`, the return value is not always a `BinaryIO`.

**Review by chunyang — APPROVED** [review]
Code LGTM, though I think one of the type signatures is wrong and there are a few places where code code be simplified with use of the standard library.

File: python/rikai/io.py (hunk 2 above)
Comment (eddyxu) [reply python/rikai/io.py:126]: Can shutil work with cloud storage?

File: python/rikai/io.py (hunk 3 above)
Comment (eddyxu) [reply python/rikai/io.py:52]: TIL

File: python/rikai/io.py (hunk 4 above)
Comment (eddyxu) [reply python/rikai/io.py:163]: Hmm, good point.

File: python/rikai/io.py (hunk 2 above)
Comment (chunyang) [reply python/rikai/io.py:126]: I think [`shutil.copyfileobj`](https://docs.python.org/3/library/shutil.html#shutil.copyfileobj) can work with any file-like object.

File: python/rikai/io.py (hunk 3 above)
Comment (eddyxu) [reply python/rikai/io.py:52]: `@functools.cache` is new from python 3.9...
We need to provide compatibility from 3.7 tho.

https://docs.python.org/3/library/functools.html

Other alternatives?

File: python/rikai/io.py (hunk 2 above)
Comment (eddyxu) [reply python/rikai/io.py:126]: `shutil.copyfileobj` ?

File: python/rikai/io.py (hunk 3 above)
Comment (eddyxu) [reply python/rikai/io.py:52]: guess we can use `lru`  🤫

**Pull request merged**